File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-4.go
func lengthOfLongestSubstring(s string) int {
	hashMap := make(map[byte]int)
	left, result := 0, 0
	for right := 0; right < len(s); right++ {
		if idx, ok := hashMap[s[right]]; ok && idx >= left {
			left = idx + 1
		}
		hashMap[s[right]] = right
		result = max(result, right-left+1)
	}
	return result
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: Data Structures & Algorithms/trapping-rain-water/submission-4.java
class Solution {
    public int trap(int[] height) {
        int result = 0, n = height.length;
        int maxLeft = 0, maxRight = 0;
        int left = 0, right = n - 1;
        while(left < right) {
            if(height[left] < height[right] ){
                maxLeft = Math.max(maxLeft, height[left]);
                result += maxLeft - height[left];
                left++;
            } else {
                maxRight = Math.max(maxRight, height[right]);
                result += maxRight - height[right];
                right--;
            }
        }
        return result;
    }
}